src/main/java/gov/cabinetoffice/gap/adminbackend/security/CheckSchemeOwnershipAspect.java

```diff
@@ -0,0 +1,140 @@
+            "applicationFormExistsDTO", this::extractGrantSchemeIdFromDto
+    );
+
+
+    private Integer fetchSchemeIdFromMethodArgs(Map<String, Object> methodArgs) {
+        for (Map.Entry<String,Object> entry : methodArgs.entrySet()) {
+            if (schemeIdExtractors.containsKey(entry.getKey())) {
+                return schemeIdExtractors.get(entry.getKey()).extractSchemeId(methodArgs.get(entry.getKey()));
+            }
+        }
+        throw new IllegalArgumentException("Unable to retrieve scheme Id from request " + methodArgs.keySet());
+    }
+
+    public Integer extractSchemeIdFromAdvert(Object value) {
+
+        if (value instanceof UUID id) {
+            GrantAdvert advert = grantAdvertRepository.findById(id).orElseThrow(NotFoundException::new);
+            return advert.getScheme().getId();
+        } else {
+            throw new IllegalArgumentException("Unable to extract grantSchemeId from Advert: " +
+                    "value is not an instance of UUID " + value);
+        }
+    }
+
+    public Integer extractSchemeIdFromApplication(Object value) {
+
+        if (value instanceof Integer id) {
+            ApplicationFormEntity application = applicationFormRepository
+                    .findById(id).orElseThrow(NotFoundException::new);
+            return application.getGrantSchemeId();
+        } else {
+            throw new IllegalArgumentException("Unable to extract grantSchemeId from Application: " +
+                    "value is not an instance of Integer " + value);
+        }
+    }
+
+    public <T> Integer extractGrantSchemeIdFromDto(T dto) {
+        try {
+            Method method = dto.getClass().getMethod("getGrantSchemeId");
+            return (Integer) method.invoke(dto);
+        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException e) {
+            throw new IllegalArgumentException("Unable to extract grantSchemeId from DTO", e);
+        }
+    }
+
+    private Integer castResponseIdToInteger(Object value) {
+        if (value instanceof Integer id) {
+            return id;
+        }
+
+        // In some cases String integers are used in the request
+        try {
+            return Integer.parseInt(value.toString());
+        } catch (NumberFormatException | NullPointerException e) {
+            throw new IllegalArgumentException("Value cannot be parsed to an integer: " + value);
+        }
+    }
+
+    private Map<String, Object> getMethodArguments(JoinPoint joinPoint) {
+        Object[] methodArgs = joinPoint.getArgs();
+        MethodSignature methodSignature = (MethodSignature) joinPoint.getSignature();
+        String[] parameterNames = methodSignature.getParameterNames();
+
+        Map<String, Object> argumentsMap = new HashMap<>();
+        int i = 0;
+        for (String parameterName : parameterNames) {
+            argumentsMap.put(parameterName, methodArgs[i++]);
+        }
+        return argumentsMap;
```

Comment:
I think this is synonymous 
```suggestion
        return IntStream.range(0, parameterNames.length)
                .boxed()
                .collect(Collectors.toMap(i -> parameterNames[i], i -> methodArgs[i]));
```